### PR TITLE
Prometheus Sink Boiler plate code for issue #1744.

### LINKED ISSUE #1744: Support Prometheus as a Sink (pull/scrape model)
Create a Sink which provides metrics to Prometheus as an API so that Prometheus can scrape Data Prepper to get metrics.

### PR DESCRIPTION
**Pull request by mallikagogoi7**
### Description
Prometheus Sink Boiler plate code for issue #1744.
 
### Issues Resolved
#1744
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    public String getUrl() {
+        return url;
+    }
+
+    public int getStatusCode() {
+        return statusCode;
+    }
+
+    public String getErrMessage() {
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java:34]: Rename: `getErrorMessage`

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        return errMessage;
+    }
+
+    @Override
+    public String toString() {
+        return "{" +
+                "url='" + url + '\'' +
+                ", statusCode=" + statusCode +
+                ", errMessage='" + errMessage + '\'' +
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java:43]: Rename `errMessage` -> `errorMessage`.

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/PrometheusSink.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+
+@DataPrepperPlugin(name = "prometheus", pluginType = Sink.class, pluginConfigurationType = PrometheusSinkConfiguration.class)
+public class PrometheusSink extends AbstractSink<Record<Event>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PrometheusSink.class);
+
+    private static final String BUCKET = "bucket";
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/PrometheusSink.java:32]: I don't think you need these two lines.

What would the Prometheus sink need a bucket for?

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/AuthTypeOptions.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+package org.opensearch.dataprepper.plugins.sink.configuration;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public enum AuthTypeOptions {
+    HTTP_BASIC("http_basic"),
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/AuthTypeOptions.java:14]: I think for consistency with other values, these should be `http-basic` and `bearer-token`.

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/ThresholdOptions.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    private static final String DEFAULT_BYTE_CAPACITY = "50mb";
+
+    @JsonProperty("event_count")
+    @Size(min = 0, max = 10000000, message = "event_count size should be between 0 and 10000000")
+    @NotNull
+    private int eventCount;
+
+    @JsonProperty("maximum_size")
+    private String maximumSize = DEFAULT_BYTE_CAPACITY;
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/ThresholdOptions.java:29]: Please use the `ByteCount` type here.

See for example: https://github.com/opensearch-project/data-prepper/blob/f8091635d2c60bdc3f9b17a9033b4c67e0fc739c/data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/configuration/ThresholdOptions.java#L49-L51

File: data-prepper-plugins/prometheus-sink/build.gradle [hunk 6]
```diff
@@ -0,0 +1,27 @@
+
+dependencies {
+    implementation project(':data-prepper-api')
+    implementation libs.armeria.core
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/build.gradle:4]: Do you need Armeria? I don't see it in use.

File: data-prepper-plugins/prometheus-sink/build.gradle [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    implementation project(path: ':data-prepper-plugins:common')
+    implementation 'com.arpnetworking.metrics:prometheus-remote-protocol:1.0.1'
+    implementation project(':data-prepper-plugins:aws-plugin-api')
+    implementation 'io.micrometer:micrometer-core'
+    implementation 'com.fasterxml.jackson.core:jackson-core'
+    implementation 'com.fasterxml.jackson.core:jackson-databind'
+    implementation 'org.apache.commons:commons-compress:1.21'
+    implementation 'joda-time:joda-time:2.11.1'
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/build.gradle:12]: Why are you using Joda? We should use Java 8 time whenever possible.

File: data-prepper-plugins/prometheus-sink/build.gradle [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    implementation project(':data-prepper-plugins:common')
+    implementation project(path: ':data-prepper-plugins:common')
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/build.gradle:14]: This is a duplicate of the line above.

File: data-prepper-plugins/prometheus-sink/build.gradle [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    implementation 'com.fasterxml.jackson.core:jackson-core'
+    implementation 'org.hibernate.validator:hibernate-validator:7.0.5.Final'
+    implementation 'com.fasterxml.jackson.dataformat:jackson-dataformat-yaml'
+    implementation 'software.amazon.awssdk:auth'
+    implementation 'org.apache.commons:commons-lang3:3.12.0'
+    implementation project(':data-prepper-plugins:failures-common')
+    implementation group: 'org.apache.httpcomponents.client5', name: 'httpclient5', version: '5.2'
```
Comment (dlvenable) [inline data-prepper-plugins/prometheus-sink/build.gradle:21]: Please use the string notation here. You are using it on all the other dependencies. 

Also, are you using this or Armeria for the client? Do we need both?

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/FailedHttpResponseInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.dataprepper.plugins.sink;
```
[the remark this answers is not in the document]
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/FailedHttpResponseInterceptor.java]: Thanks for review @dlvenable,updated the packages

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java (hunk 1 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java]: Thanks for review @dlvenable , Fixed

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java (hunk 2 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/HttpEndPointResponse.java]: Thanks for the review @dlvenable , Fixed

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/PrometheusSink.java (hunk 3 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/PrometheusSink.java]: Thanks for the review @dlvenable , Fixed

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/AuthTypeOptions.java (hunk 4 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/AuthTypeOptions.java]: Thanks for the review @dlvenable , Fixed

File: data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/ThresholdOptions.java (hunk 5 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/configuration/ThresholdOptions.java]: Thanks for the review @dlvenable , ThresholdOptions class is already same like s3-sink

File: data-prepper-plugins/prometheus-sink/build.gradle (hunk 6 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/build.gradle]: Thanks for the review @dlvenable , Fixed

File: data-prepper-plugins/prometheus-sink/build.gradle (hunk 7 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/build.gradle]: Thanks for the review @dlvenable , Fixed

File: data-prepper-plugins/prometheus-sink/build.gradle (hunk 8 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/build.gradle]: Thanks for the review @dlvenable , Fixed

File: data-prepper-plugins/prometheus-sink/build.gradle (hunk 9 above)
Comment (mallikagogoi7) [reply data-prepper-plugins/prometheus-sink/build.gradle]: Thanks for the review @dlvenable , Fixed

**Review by dlvenable — APPROVED** [review]

**Review by oeyh — APPROVED** [review]

**Pull request merged**